### PR TITLE
task/CR-1401-extended-themes-before-ls-6-can-cause-issue-with-the-attributes-when-generating-xml-files

### PR DESCRIPTION
**Pull request by marticos**
Fixed issue Jira ticket:
https://limesurvey.atlassian.net/browse/CR-1401

This Pull Request (PR) addresses two separate issues:

**1. Errors in config.xml for Imported Themes**
Problem:
Incorrect nodes in the config.xml file were causing errors. These inconsistencies may have been introduced manually by users.

Solution:

Implemented a validation check for the config.xml file during every operation involving any imported theme.
The system now identifies incorrect nodes, restores default values, and adds any missing <optgroup> elements automatically.

How to Reproduce and Verify the Fix:

Import the Provided Theme:
https://limesurvey.atlassian.net/jira/software/c/projects/MW/boards/12?assignee=712020%3A76ced84e-9120-49e6-a35b-e54369fb4980&selectedIssue=MW-1348&useStoredSettings=true

Check the config.xml File:

Incorrect Node Example:
`<cssframework type="dropdown" category="Simple options" width="12" title="Variations" parent="cssframework">
    <dropdownoptions><option data-value="css/variations/theme_apple.css" value="apple">Apple</option><option data-value="css/variations/theme_blueberry.css" value="blueberry">Blueberry</option><option data-value="css/variations/theme_grape.css" value="grape">Grape</option><option data-value="css/variations/theme_mango.css" value="mango">Mango</option></dropdownoptions>
</cssframework>`

Expected Fixed Node:
`<cssframework type="dropdown" category="Simple options" width="12" title="Variations" parent="cssframework">
            Apple<dropdownoptions><optgroup><option data-value="css/variations/theme_apple.css" value="apple">Apple</option><option data-value="css/variations/theme_blueberry.css" value="blueberry">Blueberry</option><option data-value="css/variations/theme_grape.css" value="grape">Grape</option><option data-value="css/variations/theme_mango.css" value="mango">Mango</option></optgroup></dropdownoptions>
        </cssframework>`

**2. 500 Error When Copying Surveys with Specific Theme Configurations**
Problem:
When copying a survey that uses a theme with @attributes in the template_configuration.options JSON, a PHP 500 error occurs.

Solution:

Modified the code to handle @attributes within the cssframework node, preventing the 500 error during the survey copying process.
How to Test the Fix:

Create a Test Survey:
Set the theme to "Fruity Twentythree" as the default template.
Attempt to Copy the Survey:
Try creating a new survey by copying the test survey.
Expected Outcome: The survey should copy successfully without triggering a 500 error.

If No Error Occurs:
Verify the database for a theme entry with @attributes in the template_configuration.options column:

`{"add":["css/variations/theme_apple.css","css/base.css","css/custom.css"], "remove":["survey.css", "template-core.css", "awesome-bootstrap-checkbox/awesome-bootstrap-checkbox.css", "awesome-bootstrap-checkbox/awesome-bootstrap-checkbox-rtl.css"]}	{"add":["scripts/theme.js","scripts/custom.js"], "remove":["survey.js", "template-core.js"]}	{"add":["css/print_theme.css"]}	"{"hideprivacyinfo":"off","showpopups":"1","showclearall":"off","questionhelptextposition":"top","fixnumauto":"enable","backgroundimage":"off","backgroundimagefile":".\/files\/pattern.png",
                                     "brandlogo":"off","brandlogofile":"image::theme::files\/logo.png","font":"ibm-sans",
                                     "cssframework":{"@attributes":{"type":"dropdown","category":"Simple options",
                                     "width":"12","title":"Variations","parent":"cssframework"}}}"				{"add":["pjax","moment","font-ibm-sans","font-ibm-serif"]}`

Make sure that the cssframework node includes the @attributes as shown above.

If you require help replicating the issues or verifying the fixes, please don't hesitate to reach out.

Thank you.

### CODE REVIEW

**Comment by marticos** [comment]
> Overall the code looks good, even though the problem did not reproduce for me, so I would like to ask you whether you tested it. Also, at https://github.com/LimeSurvey/LimeSurvey/pull/3792/files#diff-879ad87980b966696a7b02de0c55e0b02565637e26280383632f6c8081a9c30fR147 you have json_decode where formerly a json_encode was present. Can you explain the reason of this?

Hi @lajosarpad to reproduce the error you need a theme with an invalid or faulty config file.
Please downgrade the db of your survey site to version 623, so before applying my changes. Try to upload/import this theme with the buggy config file and you will see the 500 error.
[extends_fruity.zip](https://github.com/user-attachments/files/17116890/extends_fruity.zip)
After upgrading to version 624 and applying my changes you can try to upload it again and you will see that the theme will be correctly imported without any error and the config.xml file will be fixed.

For what concerns the json_decod is meant to fix the invalid argument error that was caught in the pipeline.
https://github.com/LimeSurvey/LimeSurvey/actions/runs/10997318082/job/30532557802

Thanks.

File: application/models/services/SurveyDeactivate.php [hunk 1]
```diff
@@ -144,7 +144,10 @@ protected function archiveToken($iSurveyID, $date, $userID, $DBDate, &$aData)
             'token',
             $DBDate,
             $aData['aSurveysettings']['tokenencryptionoptions'],
-            json_encode($aData['aSurveysettings']['attributedescriptions'])
+            json_decode(
```
[the remark this answers is not in the document]
Comment (marticos) [reply application/models/services/SurveyDeactivate.php:147]: This is done to fix the error "Invalid argument" in the pipeline.
https://github.com/LimeSurvey/LimeSurvey/actions/runs/10997318082/job/30532557802

[the remark this answers is not in the document]
Comment (lajosarpad) [reply application/models/services/SurveyDeactivate.php:147]: Yes, indeed, but we will need to make sure that we do not cause some other problems. I remember doing a similar fix for a similar problem, maybe this exact fix for this exact problem, I'm not sure and some other solution was chosen instead. Please reach out to someone who knows why it was solved the way it was back then.

**Comment by lajosarpad** [comment]
> > Overall the code looks good, even though the problem did not reproduce for me, so I would like to ask you whether you tested it. Also, at https://github.com/LimeSurvey/LimeSurvey/pull/3792/files#diff-879ad87980b966696a7b02de0c55e0b02565637e26280383632f6c8081a9c30fR147 you have json_decode where formerly a json_encode was present. Can you explain the reason of this?
> 
> Hi @lajosarpad to reproduce the error you need a theme with an invalid or faulty config file. Please downgrade the db of your survey site to version 623, so before applying my changes. Try to upload/import this theme with the buggy config file and you will see the 500 error. [extends_fruity.zip](https://github.com/user-attachments/files/17116890/extends_fruity.zip) After upgrading to version 624 and applying my changes you can try to upload it again and you will see that the theme will be correctly imported without any error and the config.xml file will be fixed.
> 
> For what concerns the json_decod is meant to fix the invalid argument error that was caught in the pipeline. https://github.com/LimeSurvey/LimeSurvey/actions/runs/10997318082/job/30532557802
> 
> Thanks.

I did try that way but with a fruity_twentythree extension, just like you described:

image

and version is 623:

image

Nevertheless, the code looks good as long as you discuss that json_decode change with the author of that line.

File: application/models/services/SurveyDeactivate.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (marticos) [reply application/models/services/SurveyDeactivate.php]: Hi Arpad, this has been restored to the previous version. I spoke with the author and he changed the method back to request a json string. Thanks.

**Review by lajosarpad — APPROVED** [review]

**Pull request merged**